python/whylogs/core/constraints/metric_constraints.py

```diff
@@ -171,20 +172,145 @@ def validate_profile(self, dataset_profile: DatasetProfileView) -> Tuple[bool, O
+                    break
+        return valid, metric_map
+
+
+class PrefixCondition:
+    def __init__(self, expression: str) -> None:
+        self._expression = expression.split()
+        self._profile = None
+        self._metric_map: Dict[str, Metric] = dict()
+
+    def _interpret(self, i: int) -> Tuple[Any, int]:
+        token = self._expression[i]
+        # ...
+        if token.startswith(":"):
+            # TODO: lexical validation
+            column_name, path = token[1:].split(":")
+            metric_name, component_name = path.split("/")  # TODO: verify this works with MultiMetric
+            try:
+                metric = self._profile.get_column(column_name).get_metric(metric_name)  # type: ignore
+                summary = metric.to_summary_dict()
+            except:  # noqa
+                raise MissingMetric(token)
+
+            metric_path = f"{column_name}/{path}"
+            self._metric_map[metric_path] = metric
+            try:
+                value = summary[component_name]
+            except:  # noqa
+                raise ValueError("Component {component_name} not found in {metric_path}")
+
+            return value, i + 1
+
+        if token == "==":
+            left, i = self._interpret(i + 1)
+            right, i = self._interpret(i)
+            return (left == right), i
+        # ...
+        raise ValueError("Unparsable expression {' '.join(self._expression)}")
+
+    def __call__(self, profile: DatasetProfileView) -> Tuple[bool, Optional[Dict[str, Metric]]]:
+        """
+        relational operators: ~ ~= == < <= > >= !=
+        boolean operators:    and or not
+        arithmetic operators: + - * / %
+
+        literals: 42 3.14 "blah"
+
+        metric component references: :column_name:metric_namespace/metric_component
+
+        returns a map of :column_name:metric_namespace -> Metric
+        """
+        self._profile = profile
+        self._metric_map = dict()
+        passes, _ = self._interpret(0)
+        if not isinstance(passes, bool):
+            raise ValueError("Contraint expression should return a boolean, got {type(passes)}")
+
+        return passes, self._metric_map
+
+
+def _make_report(
+    constraint_name: str, result: bool, with_summary: bool, metric_summary: Optional[Dict[str, Any]]
+) -> ReportResult:
+    if not result:
+        if with_summary:
+            return ReportResult(name=constraint_name, passed=0, failed=1, summary=metric_summary)
+        return ReportResult(name=constraint_name, passed=0, failed=1, summary=None)
+
+    if with_summary:
+        return ReportResult(name=constraint_name, passed=1, failed=0, summary=metric_summary)
+    return ReportResult(name=constraint_name, passed=1, failed=0, summary=None)
+
+
 class Constraints:
     column_constraints: Dict[str, Dict[str, MetricConstraint]]
-    dataset_constraints: Dict[str, MetricConstraint]
+    dataset_constraints: List[DatasetConstraint]
```

Comment:
Potential compatibility failure. `ConstraintsBuilder::add_constraint()` should maintain compatible semantics, so hopefully nobody is messing with `Constraints::dataset_constraints` directly.

Reply:
I think that's fair.